File: benchmarks/allocator/ArenaAllocator.c
// stark-bench: skip-c-windows
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>

#define ROUNDS UINT16_C(1200)
#define ITEMS_PER_ROUND UINT16_C(256)
#define SELF_TEST_ROUNDS UINT16_C(7)
#define SELF_TEST_SEED UINT64_C(19)
#define SELF_TEST_EXPECTED UINT64_C(2928019188430704367)

typedef struct {
  uint64_t left;
  uint64_t right;
} Pair;

typedef struct {
  uint8_t *data;
  size_t capacity;
  size_t offset;
} Arena;

static size_t align_up(size_t value, size_t alignment) {
  return (value + alignment - 1u) & ~(alignment - 1u);
}

static void *arena_alloc(Arena *arena, size_t size, size_t alignment) {
  size_t aligned = align_up(arena->offset, alignment);
  size_t end = aligned + size;
  if (end > arena->capacity) {
    return NULL;
  }

  arena->offset = end;
  return arena->data + aligned;
}

__attribute__((noinline)) static uint64_t run_round(uint64_t seed) {
  enum {
    PAIR_ALIGN = _Alignof(Pair),
    U64_ALIGN = _Alignof(uint64_t),
    MAX_ALIGN = PAIR_ALIGN > U64_ALIGN ? PAIR_ALIGN : U64_ALIGN
  };
  size_t capacity = align_up(sizeof(uint64_t) * ITEMS_PER_ROUND, MAX_ALIGN) +
                    (sizeof(Pair) + MAX_ALIGN) * ITEMS_PER_ROUND;
  Arena arena = {
      .data = (uint8_t *)malloc(capacity),
      .capacity = capacity,
      .offset = 0,
  };
  if (arena.data == NULL) {
    return 0;
  }

  uint64_t *values =
      (uint64_t *)arena_alloc(&arena, sizeof(uint64_t) * ITEMS_PER_ROUND,
                              _Alignof(uint64_t));
  if (values == NULL) {
    free(arena.data);
    return 0;
  }

  uint64_t checksum = seed + UINT64_C(1469598103934665603);
  for (uint16_t index = 0; index < ITEMS_PER_ROUND; index += 1) {
    uint64_t wide_index = (uint64_t)index;
    Pair *pair = (Pair *)arena_alloc(&arena, sizeof(Pair), _Alignof(Pair));
    if (pair == NULL) {
      free(arena.data);
      return 0;
    }

    pair->left = checksum + ((wide_index + UINT64_C(1)) * UINT64_C(1099511628211));
    pair->right =
        (checksum ^ (wide_index * UINT64_C(780291637))) + UINT64_C(1442695040888963407);
    uint64_t mixed =
        ((pair->left * UINT64_C(6364136223846793005)) ^ pair->right) + wide_index;
    values[index] = mixed;
    checksum = (checksum ^ mixed) * UINT64_C(1099511628211);
  }

  for (uint16_t index = 0; index < ITEMS_PER_ROUND; index += 1) {
    checksum = (checksum + values[index]) ^
               ((uint64_t)index * UINT64_C(11400714819323198485));
  }

  free(arena.data);
  return checksum;
}

__attribute__((noinline)) static uint64_t run(uint16_t rounds, uint64_t seed) {
  uint64_t checksum = seed;
  for (uint16_t round = 0; round < rounds; round += 1) {
    checksum = run_round(checksum + ((uint64_t)round * UINT64_C(7046029254386353131)));
  }

  return checksum;
}

int main(void) {
  if (run(SELF_TEST_ROUNDS, SELF_TEST_SEED) != SELF_TEST_EXPECTED) {
    return 1;
  }

  uint64_t seed = ((uint64_t)getpid()) + SELF_TEST_SEED;
  uint64_t result = run(ROUNDS, seed);
  return (result == 0 || result == seed) ? 2 : 0;
}

File: benchmarks/allocator/ArenaAllocator.rs
const ROUNDS: u16 = 1200;
const ITEMS_PER_ROUND: usize = 256;
const SELF_TEST_ROUNDS: u16 = 7;
const SELF_TEST_SEED: u64 = 19;
const SELF_TEST_EXPECTED: u64 = 2_928_019_188_430_704_367;

struct Pair {
    left: u64,
    right: u64,
}

#[inline(never)]
fn run_round(seed: u64) -> u64 {
    let mut pairs: Vec<Pair> = Vec::with_capacity(ITEMS_PER_ROUND);
    let mut values: Vec<u64> = Vec::with_capacity(ITEMS_PER_ROUND);
    let mut checksum = seed.wrapping_add(1_469_598_103_934_665_603);

    for index in 0..ITEMS_PER_ROUND {
        let wide_index = index as u64;
        pairs.push(Pair {
            left: checksum.wrapping_add((wide_index.wrapping_add(1)).wrapping_mul(1_099_511_628_211)),
            right: (checksum ^ wide_index.wrapping_mul(780_291_637))
                .wrapping_add(1_442_695_040_888_963_407),
        });

        let pair = &pairs[index];
        let mixed = (pair
            .left
            .wrapping_mul(6_364_136_223_846_793_005)
            ^ pair.right)
            .wrapping_add(wide_index);
        values.push(mixed);
        checksum = (checksum ^ mixed).wrapping_mul(1_099_511_628_211);
    }

    for (index, value) in values.iter().enumerate() {
        checksum = checksum.wrapping_add(*value)
            ^ ((index as u64).wrapping_mul(11_400_714_819_323_198_485));
    }

    checksum
}

#[inline(never)]
fn run(rounds: u16, seed: u64) -> u64 {
    let mut checksum = seed;
    for round in 0..rounds {
        checksum = run_round(
            checksum.wrapping_add((round as u64).wrapping_mul(7_046_029_254_386_353_131)),
        );
    }

    checksum
}

fn main() {
    if run(SELF_TEST_ROUNDS, SELF_TEST_SEED) != SELF_TEST_EXPECTED {
        std::process::exit(1);
    }

    let seed = (std::process::id() as u64).wrapping_add(SELF_TEST_SEED);
    let result = run(ROUNDS, seed);
    if result == 0 || result == seed {
        std::process::exit(2);
    }
}
